### ISSUE TITLE
Feature Request:  Interactive Table of admiral functions

### ISSUE THREAD

**Issue by bms63** [pharmaverse/admiral#2052]
### Feature Idea

A lot of users have requested abilities to easily search the admiral function base to find relevant functions for doing certain types of derivations, e.g. finding a function to convert a `--dtc` varibles to a `--dt` would be `derive_vars_dt()`.  

Past proposals have centered on making custom roxygen tags that classifies all of our functions.   However, this gets tricky as a lot of our functions can generalize to different derivations and different domains and continued maintenance of these tags could be time-consuming.  As we reduce our release schedule it will become hard to re-classify or introduce new ways to search the functions.

PROPOSAL: Build a stand-alone interactive table (separate from admiral repo) that allows users to search by DOMAINs or Common Variables to find recommended admiral functions.  This table would then be embedded into the admiral pkgdown site in certain vignetts (TBD). 

ADVANTAGES: We can continuously update the table without needing to do a CRAN release.

### Relevant Input

- [ ] New Repo housing infrastructure for an interactive table
- [ ] An easily updatable spreadsheet or database of all admiral functions
- The spreadsheet would have the following columns:
  - [ ] admiral or admiral extension function
  - [ ] ADaM variable(s) the function can used in
  - [ ] ADaM Domain(s) the function can used in
  - [ ] admiral Templates that function is used in
  - [ ] Relevant section of the ADaM IG if applicable.

### Relevant Output

A simple UI that interfaces with spreadsheet/database.  A user can easily search within the different columns  or in a global search.

**Comment by bms63** [comment]
@pharmaverse/admiral HI all.  **_First draft_** of building an interactive and searchable table for our function base.  I'm keen in this being developed outside of admiral so we can continuously update if needed.  Happy to hear of other ideas if you think there is a better way to do this.

Also...any takers?  It would be great to have this by the upcoming release.  However, not having it apart of admiral means we can update as we go/get feedback, etc.

**Comment by sadchla-codes** [comment]
@bms63 this sounds intriguing, out of curiosity,  when you say sinple UI, are you thinking the route of a possible interactive R Markdown with shiny elements (using shiny library) or DT package to store the TOC in a table format? Or a dashboard?

**Comment by bms63** [comment]
Hey @sadchla-codes  i think we are going to be playing around with it

@ddsjoberg is going to take this on and give us a quick PoC.  See where it goes from there!

**Comment by robertdevine** [comment]
Thanks @bms63 and @sadchla-codes, development of simple UI front-end backed by an API to tabulate admiral function results may prove helpful as the possibility for an open set of function implementations seems real for admiral. {i.e. generate tabular responses based upon queries formulated from interactive input at the UI controls - populated drop-down pickers, etc.). Plenty of excellent examples across the public internet provide interactive function documentation using similar approach. Thanks again.

**Comment by bundfussr** [comment]
I like the idea. However, I think there is room for improvement.

I would replace the "admiral or admiral extension function" column by an "admiral or admiral extension function call" column. I think that would be much more helpful because many of the admiral functions can be used for several derivations. For example `derive_vars_merged()` can be used for `EOSSTT`, `EOSDT`, `DCREAS`, `TRTSDTM`, `TRTEDTM`, ...

I would generalize the "Relevant section of the ADaM IG" column to a "Description" column. It should contain a short description of the derivation (including the relevant section of the ADaM IG if applicable).

I would add an "Id" column. Then it is easier to reference a certain row of the spreadsheet/database. For some variables like `ANL01FL` there may be several function calls available.

Furthermore, I would set up the package such that users can create their own spreadsheets/databases for company-, project-, or study-specific derivations.

In addition to the interactive table we could provide a function to pull function calls from the spreadsheet/dataset directly into an R script. Then the users would not need to copy the code manually from the table to the script.

**Comment by ddsjoberg** [comment]
Thanks for the great comments @bundfussr !


> I would replace the "admiral or admiral extension function" column by an "admiral or admiral extension function call" column. I think that would be much more helpful because many of the admiral functions can be used for several derivations. For example `derive_vars_merged()` can be used for `EOSSTT`, `EOSDT`, `DCREAS`, `TRTSDTM`, `TRTEDTM`, ...

Let's discuss! I think `derive_vars_merged()` is a special case since it's so general. If we wanted to include entries for every variable that gets merged in, it could have two primary implications: 1. making the listing comprehensive, which is great. 2. making the listing huge and more difficult to navigate.

> I would generalize the "Relevant section of the ADaM IG" column to a "Description" column. It should contain a short description of the derivation (including the relevant section of the ADaM IG if applicable).

I would think the function help file will outline the derivations (ie what's happening under the hood). If we duplicate that information in another location, it's more difficult to maintain.

> I would add an "Id" column. Then it is easier to reference a certain row of the spreadsheet/database. For some variables like `ANL01FL` there may be several function calls available.

Great idea

> Furthermore, I would set up the package such that users can create their own spreadsheets/databases for company-, project-, or study-specific derivations.

We are also exporting the data frame with all the details. Is that sufficient for what you have in mind here?

> In addition to the interactive table we could provide a function to pull function calls from the spreadsheet/dataset directly into an R script. Then the users would not need to copy the code manually from the table to the script.

Let's discuss this as well. I tend to think that copying the text will be easier than any function we could write to do this for the user?

**Comment by bundfussr** [comment]
> > I would replace the "admiral or admiral extension function" column by an "admiral or admiral extension function call" column. I think that would be much more helpful because many of the admiral functions can be used for several derivations. For example `derive_vars_merged()` can be used for `EOSSTT`, `EOSDT`, `DCREAS`, `TRTSDTM`, `TRTEDTM`, ...
> 
> Let's discuss! I think `derive_vars_merged()` is a special case since it's so general. If we wanted to include entries for every variable that gets merged in, it could have two primary implications: 1. making the listing comprehensive, which is great. 2. making the listing huge and more difficult to navigate.
> 
I think `derive_vars_merged()` is the common case. Most of the admiral functions can be used for several variables/parameters. Please have a look at the templates in admiralroche. The vast majority of derivations calls one of the general functions.

Yes, the listing will be huge. That's ADaM. I reckon just the Roche standard ADaMs have more than thousand derivations. Thus we need a good structure and a powerful search functionality.

> > I would generalize the "Relevant section of the ADaM IG" column to a "Description" column. It should contain a short description of the derivation (including the relevant section of the ADaM IG if applicable).
> 
> I would think the function help file will outline the derivations (ie what's happening under the hood). If we duplicate that information in another location, it's more difficult to maintain.

This was intended for the case that we replace the "admiral or admiral extension function" column by an "admiral or admiral extension function call" column. For example for `LASTFL` we could have "Flag last value for each patient, test, and visit, baseline observations are ignored" or for `AOCCSIFL` we could have "Most severe AE first occurrence per patient per body system". While the description of the function (`derive_var_extreme_flag()`) which is used to derive these variable just states "Add a variable flagging the first or last observation within each by group".
> 
> > Furthermore, I would set up the package such that users can create their own spreadsheets/databases for company-, project-, or study-specific derivations.
> 
> We are also exporting the data frame with all the details. Is that sufficient for what you have in mind here?

I am not sure. It depends on which functions are provided by the package.
> 
> > In addition to the interactive table we could provide a function to pull function calls from the spreadsheet/dataset directly into an R script. Then the users would not need to copy the code manually from the table to the script.
> 
> Let's discuss this as well. I tend to think that copying the text will be easier than any function we could write to do this for the user?

If it is just a single derivation, copying might be easier. I was thinking of the ADaM scripts. Some of the scripts implementing the Roche standard have more than thousand lines. This makes them hard to browse, maintain, and review. I had something like
```
adrs <- call_derivations(
  adrs,
  derivations = c("param_pd", "param_rsp", "param_bor")
```
in mind. This would pull the derivations for the PD, RSP, and BOR parameters from the database and apply them to `adrs`.

This would make the scripts much shorter. Furthermore, it would simplify the review because it would be obvious that standard derivation are used, which require less QC.

**Comment by bms63** [comment]
Location of repo: https://github.com/pharmaverse/admiraldiscovery

**Comment by bms63** [comment]
Hey @ddsjoberg 

Thanks for getting a PoC together!  

We were discussing this issue in the backlog meeting.  We were thinking that work should pause on the specific content of admiraldiscovery until we get the function re-work more stable so updates are minimal for the content of table.  However, work could still be continued on the process of getting and updating content, checking content (urls, deprecation, etc) and visualization/prettiness of the content.

I see two issues are in the https://github.com/pharmaverse/admiraldiscovery repo. I have linked them to the [Core Board](https://github.com/orgs/pharmaverse/projects/12). I also created a third issue on where this should exist in the admiral tome.

Are there other issues or ideas you would like to take from here?  If not, I think we can close this issue and continue the discussion in `admiraldiscovery`
